### PR TITLE
Undo redo buttons

### PR DESCRIPTION
**Pull request by grliszas14**
Resolves: https://github.com/audacity/audacity/issues/7549
Resolves: https://github.com/audacity/audacity/issues/7371

### CODE REVIEW

**Review by igorkorsukov — APPROVED** [review]

File: src/appshell/qml/ProjectPage/ProjectPage.qml [hunk 1]
```diff
@@ -123,6 +123,26 @@ DockPage {
...
+            closable: false
+            resizable: false
+            separatorsVisible: false
+
+            alignment: DockToolBarAlignment.Right
+            contentBottomPadding: 2
+
+            UndoRedoToolBar {
+                navigationPanel.section: root.topToolKeyNavSec
+                navigationPanel.order: 2
```
Comment (Eism) [inline src/appshell/qml/ProjectPage/ProjectPage.qml:144]: the order should be 3

File: src/trackedit/internal/au3/au3projecthistory.h [hunk 2]
```diff
@@ -19,6 +19,8 @@ class Au3ProjectHistory : public IProjectHistory
 public:
     void init() override;
 
+    muse::async::Notification isUndoRedoAvailableChanged() const override;
```
Comment (Eism) [inline src/trackedit/internal/au3/au3projecthistory.h:22]: I would suggest to move this method down as it is not the main one for this class

File: src/trackedit/internal/trackeditactionscontroller.h [hunk 3]
```diff
@@ -28,11 +29,15 @@ class TrackeditActionsController : public ITrackeditActionsController, public mu
 public:
     void init();
 
+    bool actionEnabled(const muse::actions::ActionCode& act) const override;
+    muse::async::Channel<muse::actions::ActionCode> actionEnabledChanged() const override;
+
     bool actionChecked(const muse::actions::ActionCode& actionCode) const override;
     muse::async::Channel<muse::actions::ActionCode> actionCheckedChanged() const override;
-    bool canReceiveAction(const muse::actions::ActionCode& code) const override;
+    bool canReceiveAction(const muse::actions::ActionCode& action) const override;
```
Comment (Eism) [inline src/trackedit/internal/trackeditactionscontroller.h:37]: please rename to actionCode

File: src/trackedit/internal/trackeditactionscontroller.h [hunk 4]
```diff
@@ -28,11 +29,15 @@ class TrackeditActionsController : public ITrackeditActionsController, public mu
 public:
     void init();
 
+    bool actionEnabled(const muse::actions::ActionCode& act) const override;
```
Comment (Eism) [inline src/trackedit/internal/trackeditactionscontroller.h:32]: please rename to actionCode

File: src/trackedit/internal/trackeditactionscontroller.h [hunk 5]
```diff
@@ -14,6 +14,7 @@
 #include "iinteractive.h"
 
 #include "../itrackeditactionscontroller.h"
+#include "ui/uiaction.h"
```
Comment (Eism) [inline src/trackedit/internal/trackeditactionscontroller.h:17]: I think, not needed

File: src/trackedit/internal/trackeditactionscontroller.cpp [hunk 6]
```diff
@@ -623,7 +647,17 @@ Channel<ActionCode> TrackeditActionsController::actionCheckedChanged() const
...
+    } else if (action == UNDO) {
+        return trackeditInteraction()->canUndo();
+    } else if (action == REDO) {
+        return trackeditInteraction()->canRedo();
+    }
+
+    return true;
 }
+
+
```
Comment (Eism) [inline src/trackedit/internal/trackeditactionscontroller.cpp:663]: please remove this line

File: src/projectscene/qml/Audacity/ProjectScene/toolbars/UndoRedoToolBar.qml [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+import Audacity.ProjectScene
+
+StyledToolBarView {
+    navigationPanel.name: "UndoRedoToolBar"
+    navigationPanel.accessible.name: qsTrc("projectscene", "Undo/redo toolbar")
+
+    spacing: 0
+    //! TODO AU4: apply padding when StyledToolBarView will have padding property added
+    // padding: 6
+
```
Comment (Eism) [inline src/projectscene/qml/Audacity/ProjectScene/toolbars/UndoRedoToolBar.qml:17]: Please add rowHeight: 28
In the screenshot, you can see how the toolbar overlaps the strip below:
image

File: src/projectscene/view/toolbars/undoredotoolbarmodel.cpp [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+*/
+#include "undoredotoolbarmodel.h"
+
+#include "uicomponents/view/toolbaritem.h"
+
+using namespace au::projectscene;
+using namespace muse::uicomponents;
+using namespace muse::actions;
+
+
```
Comment (Eism) [inline src/projectscene/view/toolbars/undoredotoolbarmodel.cpp:12]: please remove this line

**Review by Eism — APPROVED** [review]

**Pull request merged**